### ISSUE TITLE
No official build for ARM (aarch64) machines

### ISSUE THREAD

**Issue by griffinschreiber** [focus-editor/focus#471]
I only see x86 builds in the releases. Please correct me if I'm wrong.

**Comment by focus-editor** [comment]
It's currently not possible to build the editor for ARM because of some dependencies.

**Comment by stillerbenjamin** [comment]
#471 Failed attempt: When you're ready to compile on Apple's silicon, the following error message is returned.

# In Workspace 3 ("focus"):
/Users/Shared/Focus/modules/Input/macos.jai:810,13: Error: Too many arguments: Wanted 0, got 3.

    jai_context: Context;
    push_context jai_context {
        objc_msgSend(self, selector("windowDidResize:"), notification);

/Users/Shared/Focus/modules/Objective_C/bindings/arm64/message.jai:20,17: Info: ... Here is the procedure being called.

    objc_msgSend :: () -> void #foreign libobjc;

**Comment by griffinschreiber** [comment]
OK. Back to Doom Emacs for now I guess...

**Comment by griffinschreiber** [comment]
I'm eagerly watching this project though, it looks really cool.

**Comment by focus-editor** [comment]
If it's an apple silicon Mac, then the x64 builds should work fine there.

**Comment by griffinschreiber** [comment]
It's an aarch64 chromebook running a Debian VM lol.

**Comment by ileonte** [comment]
I've revived my experimental cross-compile branch and I've updated our nightly builds to generate `aarch64` binaries. You can find theme here: https://ileonte.dev/focus-builds/Linux/nightly/aarch64/

Caveats:
* these are **HIGHLY** experimental and **EXTREMELY UNTESTED**. The only data-point I have is that I can run these builds my M2 Macbook running Asahi, so hopefully that means other `aarch64` systems will run them too
* these will update slower than the "regular" nightlies since I have to manually rebase my branch each time something changes on `main` (which I will 100% lag behind)

**Comment by griffinschreiber** [comment]
Wow, thank you so much! I'm running the Debian 12 (Bookworm) sandbox on a Chromebook, so it's fine if something goes wrong. I'll see if I can get it working.

**Comment by griffinschreiber** [comment]
OK, so I ran it and here's the terminal output:
```
>_ ./focus-linux-aarch64-nightly-ba45b42
Could not open file "/home/hackleberry/.local/share/applications/dev.focus-editor.focus.desktop": code 2, No such file or directory
Could not open file "/home/hackleberry/.local/share/applications/dev.focus-editor.focus.desktop": code 2, No such file or directory
xkbcommon: loaded libxkbcommon.so.0
WAYLAND: loaded libwayland-client.so.0
WAYLAND: loaded libwayland-egl.so.1
WAYLAND: loaded libwayland-cursor.so.0
XCB: loaded libxcb.so.1
XCB: loaded libxcb-xkb.so.1
XCB: loaded libxcb-cursor.so.0
xkbcommon: loaded libxkbcommon-x11.so.0
xkbcommon: loaded libxkbcommon.so.0
XCB: initializing with Xlib connection
X11: loaded libX11.so.6
X11: loaded libX11-xcb.so.1
XCB: max_request_length=4194303
XCB: loaded libxcb-xinput.so.0
XCB: xcb-xkb version 1.0
GLX: loaded libGLX.so.0
Started watching external directory '/home/hackleberry/.config/focus-editor'
Could not convert relative path to absolute path: 2 No such file or directory
Couldn't start watching file /home/hackleberry/.config/focus-editor/global.focus-config because it's not found
Unable to stat file '/home/hackleberry/.local/share/focus-editor/temp/.': 22 Invalid argument
0:1(10): error: GLSL 3.30 is not supported. Supported versions are: 1.10, 1.20, and 1.00 ES
0:1(10): error: GLSL 3.30 is not supported. Supported versions are: 1.10, 1.20, and 1.00 ES
0:1(10): error: GLSL 3.30 is not supported. Supported versions are: 1.10, 1.20, and 1.00 ES
[...]
```
It also opened this window:
![image](https://github.com/user-attachments/assets/569efeae-9c3f-4d9f-9c8b-d090cee8568c)

A lot of these things are probably related to the fact that I'm using a Chromebook that I think has non-standard versions of OpenGL stuff, and I probably have done some dumb things in other places, but I hope this is helpful.

**Comment by ileonte** [comment]
Yeah, there are minimal OpenGL requirements that look like are not being met there. Unless you can somehow upgrade your drivers (mesa?) there's probably no way to fix it, we don't have any kind of fallback for older OpenGL versions.

Still, the program ran to that point so that's great I guess :)

**Comment by griffinschreiber** [comment]
OK, seems like a problem on my end not yours. I'll see what I can do.

**Comment by griffinschreiber** [comment]
So the program is saying that my GLSL version is below the minimum required version. From what I can tell, if the program doesn't actually need features from GLSL 3.30 then I can tell it I'm using 3.30 but not actually use it, which might cause weird stuff or undefined behaviour if it actually needs features from 3.30.
Anyway, the environment variable is `MESA_GL_VERSION_OVERRIDE`, and you can do:
`export MESA_GL_VERSION_OVERRIDE=3.3`
for anyone in the future with the same problem.
**NOTE:** I have not tried this and it may cause weird or undefined behaviour.

**Comment by ileonte** [comment]
As I said in another comment an official ARM64 build is not going to gappen any time soon. I will maintain my `aarch64` branch as close to `main` as possible but for the foreseeable future this will only generate (very) rare "nightly" builds.

If you have any problems with the `aarch64` nightlies feel free to drop by our Discord server and ping me in the `linux` channel.

**Comment by griffinschreiber** [comment]
Ok thx. My computer broke so now I'm on a Thinkpad (x64) so I'm good.
